### PR TITLE
add table regex filter for canal

### PR DESCRIPTION
**Pull request by shabicheng**
add include and exclude regex filter for canal
add filter cache for better match performance

### CODE REVIEW

File: canal/dump.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ func (h *dumpParseHandler) Data(db string, table string, values []string) error
 
 	tableInfo, err := h.c.GetTable(db, table)
 	if err != nil {
+		if errors.Trace(err) == ErrExcludedTable || errors.Trace(err) == schema.ErrTableNotExist || errors.Trace(err) == schema.ErrMissingTableMeta {
```
Comment (siddontang) [inline canal/dump.go]:
```
e := errors.Trace(err) 
if e == Err || e == Err || e == {
}
```

File: canal/sync.go [hunk 2]
```diff
@@ -77,8 +77,8 @@ func (c *Canal) runSyncBinlog() error {
 		case *replication.RowsEvent:
 			// we only focus row based event
 			err = c.handleRowsEvent(ev)
-			if err != nil && (errors.Cause(err) != schema.ErrTableNotExist && errors.Cause(err) != schema.ErrMissingTableMeta) {
-				// if error is not ErrTableNotExist or ErrMissingTableMeta, stop canal
+			if err != nil && (errors.Cause(err) != ErrExcludedTable && errors.Cause(err) != schema.ErrTableNotExist && errors.Cause(err) != schema.ErrMissingTableMeta) {
```
Comment (siddontang) [inline canal/sync.go]:
```
e := errors.Trace(err) 
// use e here
```

Comment (siddontang) [reply canal/sync.go]: Btw, the condition line is too long, you can use multi-lines.

File: canal/canal.go [hunk 3]
```diff
@@ -197,8 +218,40 @@ func (c *Canal) Ctx() context.Context {
 	return c.ctx
 }
 
+func (c *Canal) checkTableMatch(key string) bool {
+	// no filter, return true
+	if c.tableMatchCache == nil {
+		return true
+	}
+	// cache hit
+	if rst, ok := c.tableMatchCache[key]; ok {
```
Comment (siddontang) [inline canal/canal.go]: please use a lock to operate the cache map.

File: canal/config.go [hunk 4]
```diff
@@ -46,6 +46,16 @@ type Config struct {
...
 	ReadTimeout     time.Duration `toml:"read_timeout"`
 
+	// IncludeTableRegex or ExcludeTableRegex should contain database name
+	// Only a table match IncludeTableRegex and dismatch ExcludeTableRegex will be processed
+	// eg, IncludeTableRegex : ".*\\.canal", ExcludeTableRegex : "mysql\\..*"
+	//     this will include all database's 'canal' table, except database 'mysql'
+	// If IncludeTableRegex is empty, canal will not check IncludeTableRegex
+	// If ExcludeTableRegex is empty, canal will not check ExcludeTableRegex
+	// Default IncludeTableRegex and ExcludeTableRegex are empty, this will include all tables
+	IncludeTableRegex string `toml:include_table_regex`
```
Comment (siddontang) [inline canal/config.go]: how to support many different tables? 

I think this should be a `[]string`

**Comment by siddontang** [comment]
please add some tests to cover included and excluded tables.

File: canal/dump.go (hunk 1 above)
Comment (shabicheng) [reply canal/dump.go]: done

File: canal/sync.go (hunk 2 above)
Comment (shabicheng) [reply canal/sync.go]: done

File: canal/config.go (hunk 4 above)
Comment (shabicheng) [reply canal/config.go]: done

File: canal/canal.go (hunk 3 above)
Comment (shabicheng) [reply canal/canal.go]: OK, because tableMatchCache only be used in GetTable, so I use tableLock instead,.

File: canal/canal.go [hunk 5]
```diff
@@ -78,6 +84,33 @@ func NewCanal(cfg *Config) (*Canal, error) {
 		return nil, errors.Trace(err)
 	}
 
+	// init table filter
+	if c.cfg.IncludeTableRegex != nil && len(c.cfg.IncludeTableRegex) > 0 {
```
Comment (siddontang) [inline canal/canal.go]: no need to check nil here.

Comment (siddontang) [reply canal/canal.go]:
```
if n := len(c.fg.IncludeTableRegexp); n > 0 {

}
```

File: canal/canal.go [hunk 6]
```diff
@@ -197,8 +230,52 @@ func (c *Canal) Ctx() context.Context {
 	return c.ctx
 }
 
+func (c *Canal) checkTableMatch(key string) bool {
+	// no filter, return true
+	if c.tableMatchCache == nil {
+		return true
+	}
+
+	c.tableLock.Lock()
```
Comment (siddontang) [inline canal/canal.go]: the lock scope is huge, just lock when you really need to operate the cache.

Btw, use read lock for reading cache and write lock for writing cache.

File: canal/config.go [hunk 7]
```diff
@@ -46,6 +46,16 @@ type Config struct {
 	HeartbeatPeriod time.Duration `toml:"heartbeat_period"`
 	ReadTimeout     time.Duration `toml:"read_timeout"`
 
+	// IncludeTableRegex or ExcludeTableRegex should contain database name
+	// Only a table match IncludeTableRegex and dismatch ExcludeTableRegex will be processed
+	// eg, IncludeTableRegex : [".*\\.canal"], ExcludeTableRegex : ["mysql\\..*"]
+	//     this will include all database's 'canal' table, except database 'mysql'
+	// If IncludeTableRegex is empty, canal will not check IncludeTableRegex
```
Comment (siddontang) [inline canal/config.go]: remove these two If here

File: canal/config.go [hunk 8]
```diff
@@ -46,6 +46,16 @@ type Config struct {
 	HeartbeatPeriod time.Duration `toml:"heartbeat_period"`
 	ReadTimeout     time.Duration `toml:"read_timeout"`
 
+	// IncludeTableRegex or ExcludeTableRegex should contain database name
+	// Only a table match IncludeTableRegex and dismatch ExcludeTableRegex will be processed
```
Comment (siddontang) [inline canal/config.go]:
```
Only a table which matches xxx and mismatches xxx 
```

File: canal/canal.go (hunk 5 above)
Comment (shabicheng) [reply canal/canal.go]: sorry,  not familiar to go

File: canal/canal.go (hunk 6 above)
Comment (shabicheng) [reply canal/canal.go]: done

File: canal/config.go (hunk 8 above)
Comment (shabicheng) [reply canal/config.go]: done

File: canal/canal.go (hunk 5 above)
Comment (siddontang) [reply canal/canal.go]:
```
if n := len(c.fg.IncludeTableRegexp); n > 0 {
c.includeTableRegex = make([]*regexp.Regexp, n)
}
```

File: canal/canal.go [hunk 9]
```diff
@@ -197,8 +230,58 @@ func (c *Canal) Ctx() context.Context {
...
 }
 
+func (c *Canal) checkTableMatch(key string) bool {
+	// no filter, return true
+	if c.tableMatchCache == nil {
+		return true
+	}
+
+	c.tableLock.RLock()
+	if rst, ok := c.tableMatchCache[key]; ok {
```
Comment (siddontang) [inline canal/canal.go]:
```
c.tableLock.RLock()
rst, ok := xxx
c.tableLock.Runlock()
if ok {
     return rst
}
```

Comment (shabicheng) [reply canal/canal.go]: done

File: canal/config.go (hunk 7 above)
Comment (shabicheng) [reply canal/config.go]: done

File: canal/canal.go [hunk 10]
```diff
@@ -197,8 +230,58 @@ func (c *Canal) Ctx() context.Context {
...
+		// cache hit
+		c.tableLock.RUnlock()
+		return rst
+	}
+	c.tableLock.RUnlock()
+	// check include
+	if c.includeTableRegex != nil {
```
Comment (siddontang) [inline canal/canal.go]:
```
matchFlag = true

if include {
    matchFlag = false
    if matched {
       matchFlag = true
       break
    }
}

if !matchFlag && exclude {
    if matched {
        matchFlag = false
        break
    }
}

lock
cache = matchFlag
unlock
```

So you just only need using lock in one place.

Comment (siddontang) [reply canal/canal.go]: here

File: canal/canal.go [hunk 11]
```diff
@@ -78,6 +84,33 @@ func NewCanal(cfg *Config) (*Canal, error) {
 		return nil, errors.Trace(err)
 	}
 
+	// init table filter
+	if len(c.cfg.IncludeTableRegex) > 0 {
+		c.includeTableRegex = make([]*regexp.Regexp, len(c.cfg.IncludeTableRegex), len(c.cfg.IncludeTableRegex))
```
Comment (siddontang) [inline canal/canal.go]:
```
if n := len(); n > 0 {
c.includeTableRegex = make([]*regexp.Regexp,n)
}
```

**Comment by shabicheng** [comment]
add a case in canal_test

File: canal/canal.go [hunk 12]
```diff
@@ -197,8 +230,58 @@ func (c *Canal) Ctx() context.Context {
...
+
+	c.tableLock.RLock()
+	rst, ok := c.tableMatchCache[key]
+	c.tableLock.RUnlock()
+	if ok {
+		// cache hit
+		return rst
+	}
+	// check include
+	if c.includeTableRegex != nil {
```
Comment (siddontang) [inline canal/canal.go]:
```
matchFlag = true

if include {
    matchFlag = false
    if matched {
       matchFlag = true
       break
    }
}

if !matchFlag && exclude {
    if matched {
        matchFlag = false
        break
    }
}

lock
cache = matchFlag
unlock
```

Then you can use lock in one place.

Comment (shabicheng) [reply canal/canal.go]: cool

**Comment by siddontang** [comment]
LGTM

**Pull request merged**